src/lib.c

```diff
@@ -227,6 +237,250 @@ efi_variables_supported(void)
+		strncpy(guid_buf, guid_name, guid_len - 1);
+
+		rc = text_to_guid(guid_buf, guid);
+		if (rc < 0)
+			goto bad_name;
+	}
+
+	char *name_buf = NULL;
+	int name_len;
+	name_len = strlen(guid_name + name_pos) + 1;
+	name_buf = calloc(1, name_len);
+	if (!name_buf) {
+		fprintf(stderr, "efivar: %m\n");
+		exit(1);
+	}
+	strcpy(name_buf, guid_name + name_pos);
+	*name = name_buf;
+}
+
+void PUBLIC
+efi_save_variable_data(efi_variable_t *var, char *outfile, bool dmpstore)
+{
+	FILE *out = NULL;
+	ssize_t sz;
+	uint8_t *data = NULL;
+	size_t datasz = 0;
+	ssize_t (*export)(efi_variable_t *var, uint8_t *data, size_t size) =
+		dmpstore ? efi_variable_export_dmpstore : efi_variable_export;
+
+	out = fopen(outfile, "w");
+	if (!out)
+		err(1, "Could not open \"%s\" for writing", outfile);
+
+	sz = export(var, data, datasz);
+	data = calloc(sz, 1);
+	if (!data)
+		err(1, "Could not allocate memory");
+	datasz = sz;
+
+	sz = export(var, data, datasz);
+	if (sz < 0)
+		err(1, "Could not format data");
+	datasz = sz;
+
+	sz = fwrite(data, 1, datasz, out);
+	if (sz < (ssize_t)datasz)
+		err(1, "Could not write to \"%s\"", outfile);
+
+	fflush(out);
+	fclose(out);
+}
+
+void PUBLIC
+efi_save_variable(char *guid_name, char *outfile, bool dmpstore)
+{
+	efi_guid_t guid = efi_guid_empty;
+	char *name = NULL;
+	int rc;
+
+	uint8_t *data = NULL;
+	size_t data_size = 0;
+	uint32_t attributes = 7;
+	efi_variable_t *var;
+
+	efi_parse_name(guid_name, &name, &guid);
+	if (!name || efi_guid_is_empty(&guid)) {
+		fprintf(stderr, "efivar: could not parse variable name.\n");
+		show_errors();
+		exit(1);
+	}
+
+	errno = 0;
+	rc = efi_get_variable(guid, name, &data, &data_size, &attributes);
+	if (rc < 0) {
+		fprintf(stderr, "efivar: show variable: %m\n");
+		show_errors();
+		exit(1);
+	}
+
+	var = efi_variable_alloc();
+	if (!var) {
+		fprintf(stderr, "efivar: could not allocate variable storage.\n");
+		show_errors();
+		exit(1);
+	}
+
+	efi_variable_set_name(var, (unsigned char *)name);
+	efi_variable_set_guid(var, &guid);
+	efi_variable_set_attributes(var, attributes);
+	efi_variable_set_data(var, data, data_size);
+
+	efi_save_variable_data(var, outfile, dmpstore);
+
+	efi_variable_free(var, false);
+
+	free(name);
+	if (data)
+		free(data);
+}
+
+static int
+get_esp_filename(char **esp_filename)
+{
+	char *dst_name = NULL;
+	efi_guid_t dst_guid = efi_guid_empty;
+	size_t filename_size;
+	uint32_t filename_attr;
+	uint8_t *filename = NULL;
+	int rc = 0;
+
+	efi_parse_name(EFI_RTSTORAGEVOLATILE, &dst_name, &dst_guid);
+
+	if (dst_name && !efi_guid_is_empty(&dst_guid)) {
+		rc = efi_get_variable(dst_guid, dst_name, &filename, &filename_size, &filename_attr);
+		if (rc < 0) {
+			fprintf(stderr, "efivar: Error %d getting filename from RTStorageVolatile\n", rc);
+			return rc;
+		}
+
+		printf("efivar: RSTStorageVolatile found with %s\n", (char *)filename);
+	}
+
+	*esp_filename = (char *)filename;
+
+	free(dst_name);
+	free(filename);
```

Comment:
you can't free filename here, you are using its address

Reply:
Yes, you're right. Removed. Thanks!